File: tests/wasm/test_pfe.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('PFE partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.pfe_js(close, 10, 5);
    assert.strictEqual(result.length, close.length);
});

test('PFE accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.pfe;

    const result = wasm.pfe_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.smoothing
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "PFE last 5 values mismatch"
    );


    for (let i = 0; i < expected.defaultParams.period; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }




});

test('PFE default candles', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.pfe_js(close, 10, 5);
    assert.strictEqual(result.length, close.length);
});

test('PFE zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.pfe_js(inputData, 0, 5);
    }, /Invalid period/);
});

test('PFE period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.pfe_js(dataSmall, 10, 2);
    }, /Invalid period/);
});

test('PFE very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.pfe_js(singlePoint, 10, 2);
    }, /Invalid period|Not enough valid data/);
});

test('PFE empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.pfe_js(empty, 10, 5);
    }, /Input data slice is empty/);
});

test('PFE invalid smoothing', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);


    assert.throws(() => {
        wasm.pfe_js(data, 2, 0);
    }, /Invalid smoothing/);
});

test('PFE reinput', () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.pfe;


    const firstResult = wasm.pfe_js(close, 10, 5);
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.pfe_js(firstResult, 10, 5);
    assert.strictEqual(secondResult.length, firstResult.length);



    let firstNonNaN = -1;
    for (let i = 0; i < secondResult.length; i++) {
        if (!isNaN(secondResult[i])) {
            firstNonNaN = i;
            break;
        }
    }
    assert(firstNonNaN >= 20, `Expected extended warmup, but first non-NaN at ${firstNonNaN}`);
});

test('PFE NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.pfe_js(close, 10, 5);
    assert.strictEqual(result.length, close.length);


    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }


    assertAllNaN(result.slice(0, 10), "Expected NaN in warmup period");
});

test('PFE all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.pfe_js(allNaN, 10, 5);
    }, /All values are NaN/);
});

test('PFE batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.pfe_batch(close, {
        period_range: [10, 10, 0],
        smoothing_range: [5, 5, 0]
    });


    const singleResult = wasm.pfe_js(close, 10, 5);

    assert.strictEqual(batchResult.values.length, singleResult.length);

    assertArrayClose(batchResult.values, singleResult, 1e-9, "Batch vs single mismatch");
});

test('PFE batch multiple periods', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.pfe_batch(close, {
        period_range: [10, 18, 4],
        smoothing_range: [5, 5, 0]
    });


    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);


    const periods = [10, 14, 18];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);

        const singleResult = wasm.pfe_js(close, periods[i], 5);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
    }
});

test('PFE batch metadata from result', () => {

    const close = new Float64Array(30);
    close.fill(100);

    const result = wasm.pfe_batch(close, {
        period_range: [10, 20, 5],
        smoothing_range: [5, 10, 5]
    });


    assert.strictEqual(result.combos.length, 6);


    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[0].smoothing, 5);


    assert.strictEqual(result.combos[5].period, 20);
    assert.strictEqual(result.combos[5].smoothing, 10);
});

test('PFE batch full parameter sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const batchResult = wasm.pfe_batch(close, {
        period_range: [10, 14, 4],
        smoothing_range: [5, 10, 5]
    });


    assert.strictEqual(batchResult.combos.length, 4);
    assert.strictEqual(batchResult.rows, 4);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.values.length, 4 * 50);


    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const period = batchResult.combos[combo].period;
        const smoothing = batchResult.combos[combo].smoothing;

        const rowStart = combo * 50;
        const rowData = batchResult.values.slice(rowStart, rowStart + 50);


        for (let i = 0; i < period; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }


        for (let i = period; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('PFE batch edge cases', () => {

    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);


    const singleBatch = wasm.pfe_batch(close, {
        period_range: [5, 5, 1],
        smoothing_range: [3, 3, 1]
    });

    assert.strictEqual(singleBatch.values.length, 12);
    assert.strictEqual(singleBatch.combos.length, 1);


    const largeBatch = wasm.pfe_batch(close, {
        period_range: [5, 7, 10],
        smoothing_range: [3, 3, 0]
    });


    assert.strictEqual(largeBatch.values.length, 12);
    assert.strictEqual(largeBatch.combos.length, 1);


    assert.throws(() => {
        wasm.pfe_batch(new Float64Array([]), {
            period_range: [10, 10, 0],
            smoothing_range: [5, 5, 0]
        });
    }, /All values are NaN|Empty/);
});


test('PFE zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
    const period = 5;
    const smoothing = 3;


    const ptr = wasm.pfe_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');


    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );


    memView.set(data);


    try {
        wasm.pfe_into(ptr, ptr, data.length, period, smoothing);


        const regularResult = wasm.pfe_js(data, period, smoothing);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue;
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {

        wasm.pfe_free(ptr, data.length);
    }
});

test('PFE zero-copy with large dataset', () => {
    const size = 10000;
    const data = new Float64Array(size);

    for (let i = 0; i < size; i++) {
        data[i] = 100 + i * 0.01 + Math.sin(i * 0.1) * 5;
    }

    const ptr = wasm.pfe_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');

    try {
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        memView.set(data);

        wasm.pfe_into(ptr, ptr, size, 10, 5);


        const memView2 = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);


        for (let i = 0; i < 10; i++) {
            assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
        }


        for (let i = 10; i < Math.min(100, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }


        for (let i = 10; i < Math.min(100, size); i++) {
            assert(memView2[i] >= -100 && memView2[i] <= 100,
                   `PFE value ${memView2[i]} at index ${i} out of bounds [-100, 100]`);
        }
    } finally {
        wasm.pfe_free(ptr, size);
    }
});

test('PFE zero-copy error handling', () => {

    assert.throws(() => {
        wasm.pfe_into(0, 0, 10, 10, 5);
    }, /null pointer/i);


    const ptr = wasm.pfe_alloc(10);
    try {

        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, 10);
        for (let i = 0; i < 10; i++) {
            memView[i] = i + 1.0;
        }


        assert.throws(() => {
            wasm.pfe_into(ptr, ptr, 10, 0, 5);
        }, /Invalid period/);


        assert.throws(() => {
            wasm.pfe_into(ptr, ptr, 10, 5, 0);
        }, /Invalid smoothing/);
    } finally {
        wasm.pfe_free(ptr, 10);
    }
});

test('PFE zero-copy memory management', () => {

    const sizes = [100, 1000, 10000];

    for (const size of sizes) {
        const ptr = wasm.pfe_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);


        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 2.5;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 2.5, `Memory corruption at index ${i}`);
        }


        wasm.pfe_free(ptr, size);
    }
});

test('PFE batch - new ergonomic API with single parameter', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const result = wasm.pfe_batch(close, {
        period_range: [10, 10, 0],
        smoothing_range: [5, 5, 0]
    });


    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, close.length);


    const combo = result.combos[0];
    assert.strictEqual(combo.period, 10);
    assert.strictEqual(combo.smoothing, 5);


    const oldResult = wasm.pfe_js(close, 10, 5);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(result.values[i])) {
            continue;
        }
        assert(Math.abs(oldResult[i] - result.values[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('PFE batch - new API with multiple parameters', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const result = wasm.pfe_batch(close, {
        period_range: [10, 14, 4],
        smoothing_range: [5, 10, 5]
    });


    assert.strictEqual(result.rows, 4);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 4);
    assert.strictEqual(result.values.length, 200);


    const expectedCombos = [
        { period: 10, smoothing: 5 },
        { period: 10, smoothing: 10 },
        { period: 14, smoothing: 5 },
        { period: 14, smoothing: 10 }
    ];

    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.combos[i].period, expectedCombos[i].period);
        assert.strictEqual(result.combos[i].smoothing, expectedCombos[i].smoothing);
    }


    const secondRow = result.values.slice(result.cols, 2 * result.cols);
    assert.strictEqual(secondRow.length, 50);


    const oldResult = wasm.pfe_js(close, 10, 5);
    const firstRow = result.values.slice(0, result.cols);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(firstRow[i])) {
            continue;
        }
        assert(Math.abs(oldResult[i] - firstRow[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('PFE batch - new API error handling', () => {
    const close = new Float64Array(testData.close.slice(0, 10));


    assert.throws(() => {
        wasm.pfe_batch(close, {
            period_range: [10, 10],
            smoothing_range: [5, 5, 0]
        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.pfe_batch(close, {
            period_range: [10, 10, 0]

        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.pfe_batch(close, {
            period_range: "invalid",
            smoothing_range: [5, 5, 0]
        });
    }, /Invalid config/);
});


test('PFE SIMD128 consistency', () => {


    const testCases = [
        { size: 20, period: 5 },
        { size: 100, period: 10 },
        { size: 1000, period: 20 }
    ];

    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);

        for (let i = 0; i < testCase.size; i++) {
            data[i] = 100 + i * 0.5 + Math.sin(i * 0.2) * 2;
        }

        const result = wasm.pfe_js(data, testCase.period, 5);


        assert.strictEqual(result.length, data.length);


        for (let i = 0; i < testCase.period; i++) {
            assert(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }


        let countAfterWarmup = 0;
        for (let i = testCase.period; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);

            assert(result[i] >= -100 && result[i] <= 100,
                   `PFE value ${result[i]} out of bounds at index ${i}`);
            countAfterWarmup++;
        }

        assert(countAfterWarmup > 0, 'Should have values after warmup');
    }
});


test('PFE batch_into zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    const combos = 2 * 2;
    const totalSize = data.length * combos;


    const inPtr = wasm.pfe_alloc(data.length);
    const outPtr = wasm.pfe_alloc(totalSize);

    try {

        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, data.length);
        inView.set(data);


        const numCombos = wasm.pfe_batch_into(
            inPtr, outPtr, data.length,
            5, 7, 2,
            3, 5, 2
        );

        assert.strictEqual(numCombos, combos, 'Should return correct number of combinations');


        const outView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, totalSize);


        const firstRow = Array.from(outView.slice(0, data.length));
        const expected = wasm.pfe_js(data, 5, 3);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(expected[i]) && isNaN(firstRow[i])) continue;
            assert(Math.abs(firstRow[i] - expected[i]) < 1e-10,
                   `Mismatch at index ${i}: got ${firstRow[i]}, expected ${expected[i]}`);
        }
    } finally {
        wasm.pfe_free(inPtr, data.length);
        wasm.pfe_free(outPtr, totalSize);
    }
});

test.after(() => {
    console.log('PFE WASM tests completed');
});

File: tests/python/test_pfe.py
"""
Python binding tests for PFE indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestPfe:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_pfe_partial_params(self, test_data):
        """Test PFE with partial parameters (None values) - mirrors check_pfe_partial_params"""
        close = test_data['close']


        result = ta_indicators.pfe(close, 10, 5)
        assert len(result) == len(close)

    def test_pfe_accuracy(self, test_data):
        """Test PFE matches expected values from Rust tests - mirrors check_pfe_accuracy"""
        close = test_data['close']

        result = ta_indicators.pfe(close, period=10, smoothing=5)

        assert len(result) == len(close)


        expected_last_five = [-13.03562252, -11.93979855, -9.94609862, -9.73372410, -14.88374798]
        assert_close(
            result[-5:],
            expected_last_five,
            rtol=1e-8,
            msg="PFE last 5 values mismatch"
        )


        for i in range(9):
            assert np.isnan(result[i]), f"Expected NaN at index {i} during warmup"

    def test_pfe_default_candles(self, test_data):
        """Test PFE with default parameters - mirrors check_pfe_default_candles"""
        close = test_data['close']


        result = ta_indicators.pfe(close, 10, 5)
        assert len(result) == len(close)

    def test_pfe_zero_period(self):
        """Test PFE fails with zero period - mirrors check_pfe_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.pfe(input_data, period=0, smoothing=5)

    def test_pfe_period_exceeds_length(self):
        """Test PFE fails when period exceeds data length - mirrors check_pfe_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.pfe(data_small, period=10, smoothing=2)

    def test_pfe_very_small_dataset(self):
        """Test PFE fails with insufficient data - mirrors check_pfe_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.pfe(single_point, period=10, smoothing=2)

    def test_pfe_zero_smoothing(self):
        """Test PFE fails with zero smoothing"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])

        with pytest.raises(ValueError, match="Invalid smoothing"):
            ta_indicators.pfe(data, period=2, smoothing=0)

    def test_pfe_reinput(self, test_data):
        """Test PFE applied twice (re-input) - mirrors check_pfe_reinput"""
        close = test_data['close']


        first_result = ta_indicators.pfe(close, period=10, smoothing=5)
        assert len(first_result) == len(close)


        second_result = ta_indicators.pfe(first_result, period=10, smoothing=5)
        assert len(second_result) == len(first_result)


        for i in range(20, len(second_result)):
            assert not np.isnan(second_result[i]), f"Expected value after warmup, but found NaN at index {i}"

    def test_pfe_nan_handling(self, test_data):
        """Test PFE handles NaN values correctly - mirrors check_pfe_nan_handling"""
        close = test_data['close']

        result = ta_indicators.pfe(close, period=10, smoothing=5)
        assert len(result) == len(close)


        if len(result) > 240:
            for i in range(240, len(result)):
                assert not np.isnan(result[i]), f"Found unexpected NaN at index {i}"

    def test_pfe_all_nan_input(self):
        """Test PFE with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.pfe(all_nan, period=10, smoothing=5)

    def test_pfe_empty_input(self):
        """Test PFE fails with empty input"""
        empty = np.array([])

        with pytest.raises(ValueError):
            ta_indicators.pfe(empty, period=10, smoothing=5)

    def test_pfe_batch_single_parameter_set(self, test_data):
        """Test batch with single parameter combination"""
        close = test_data['close']


        batch_result = ta_indicators.pfe_batch(
            close,
            period_range=(10, 10, 0),
            smoothing_range=(5, 5, 0)
        )


        single_result = ta_indicators.pfe(close, 10, 5)

        assert batch_result['values'].shape == (1, len(close))
        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="Batch vs single mismatch"
        )


        assert len(batch_result['periods']) == 1
        assert batch_result['periods'][0] == 10
        assert len(batch_result['smoothings']) == 1
        assert batch_result['smoothings'][0] == 5

    def test_pfe_batch_multiple_parameters(self, test_data):
        """Test batch with multiple parameter values"""
        close = test_data['close'][:100]


        batch_result = ta_indicators.pfe_batch(
            close,
            period_range=(10, 12, 2),
            smoothing_range=(5, 7, 2)
        )


        assert batch_result['values'].shape == (4, 100)


        expected_periods = [10, 10, 12, 12]
        expected_smoothings = [5, 7, 5, 7]

        assert list(batch_result['periods']) == expected_periods
        assert list(batch_result['smoothings']) == expected_smoothings


        single_result = ta_indicators.pfe(close, 10, 5)
        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="First batch row mismatch"
        )

    def test_pfe_batch_edge_cases(self):
        """Test edge cases for batch processing"""
        close = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], dtype=np.float64)


        single_batch = ta_indicators.pfe_batch(
            close,
            period_range=(3, 3, 1),
            smoothing_range=(2, 2, 1)
        )

        assert single_batch['values'].shape == (1, 10)


        large_batch = ta_indicators.pfe_batch(
            close,
            period_range=(3, 5, 10),
            smoothing_range=(2, 2, 0)
        )


        assert large_batch['values'].shape == (1, 10)
        assert large_batch['periods'][0] == 3

    def test_pfe_streaming(self, test_data):
        """Test PFE streaming functionality"""
        close = test_data['close']


        stream = ta_indicators.PfeStream(period=10, smoothing=5)


        stream_results = []
        for price in close:
            result = stream.update(price)
            stream_results.append(result if result is not None else np.nan)


        batch_result = ta_indicators.pfe(close, period=10, smoothing=5)


        assert_close(
            stream_results,
            batch_result,
            rtol=1e-9,
            msg="Streaming vs batch mismatch"
        )

    def test_pfe_kernel_parameter(self, test_data):
        """Test PFE with different kernel parameters"""
        close = test_data['close'][:100]


        kernels = [None, 'scalar', 'avx2', 'avx512', 'auto']

        for kernel in kernels:
            try:
                if kernel:
                    result = ta_indicators.pfe(close, 10, 5, kernel=kernel)
                else:
                    result = ta_indicators.pfe(close, 10, 5)
                assert len(result) == len(close)
            except ValueError as e:

                msg = str(e)
                if (
                    "Unsupported kernel" not in msg
                    and "not available on this CPU" not in msg
                    and "not compiled in this build" not in msg
                ):
                    raise


if __name__ == "__main__":
    pytest.main([__file__])
